### ISSUE TITLE
no override found for vtkActor for visualizing normal (PCL1.8)

### ISSUE THREAD

**Issue by vincint810923** [PointCloudLibrary/pcl#1553]
visualizing the point cloud is working, but normal.

```
Generic Warning: In C:\Build\3rdParty\x64\VTK-7.0.0\Rendering\Core\vtkActor.cxx, line 43
Error: no override found for 'vtkActor'.
```

**Comment by SergioRAgostinho** [comment]
I'm sorry, could you please be more explicit, or even better, post a code snippet exemplifying your issue?

**Comment by pclbel** [comment]
I have the same error when I tried to visualize normals like in the "normal_estimation_using_integral_images"-tutorial. 
I have no problem on visualizing a normal point cloud, but when I try to visualize normals with the addPointCloudNormals from the pcl::visualization::PCLVisualizer class with the same pointcloud but the computed normals in addition I get a
"Access Violation"-Exception. 
When I debug, it seems to happen in the vtkSmartPointer line 117
when trying to call statement "return vtkSmartPointer<T>(T::New(), NoReference())" in the 6th execution.

In addition to that, the vtkoutputWindow has the output "Generic Warning: In C:\Build\3rdParty\x64\VTK-7.0.0\Rendering\Core\vtkActor.cxx, line 43
Error: no override found for 'vtkActor'." which is exactly the same like the one in post before.

I have installed the PCL 1.8.0 All-in-one Installer MSVC2015 x64 which can be downloaded from http://unanancyowen.com/en/pcl18/. I am on a Win10 machine with Visual Studion 2015. Everything else works as expected.

Any ideas?

**Comment by SergioRAgostinho** [comment]
@UnaNancyOwen  can you reproduce this with HEAD?

**Comment by UnaNancyOwen** [comment]
First of all, I tried the tutorial "Normal Estimation Using Integral Images" using PCL 1.8.0 that installed by PCL 1.8.0 All-in-one Installer. But, It works correctly in my environment. I have not tried it yet, but I think that it will probably work with HEAD too.

![2017-06-28_00h24_15](https://user-images.githubusercontent.com/816705/27595632-322c01f0-5b98-11e7-83e5-1bb370f78b35.png)

@pclbel Are you using CMake to generate a your project?

**Comment by pclbel** [comment]
@UnaNancyOwen No I didnt use CMake, I configured linker dependencies directly via Viusal Studio 2015.
It would have been better if I had examined it with cmake in addition.

But now I checked it again, I cant reproduce the issue any more! I remember I changed a lot the linker dependecies configuration (also vtk) during the last months. Maybe that was the reason for the error/warning, but could also be a different pointcloud used. Dont know.
So to sum up - I cant reproduce the error anymore!

**Comment by SergioRAgostinho** [comment]
Ok. Reopen if needed. Thanks for the help @UnaNancyOwen

**Comment by jolosimeon** [comment]
May I ask what changes did you do? I am having the same error now.

**Comment by pclbel** [comment]
@jolosimeon I couldnt reproduce the error with my projects built with cmake.
But as I heard you have the same problem, I examined again my project which I built manually with MSVS2015. In the project built with MSVS2015 (without cmake) I could reproduce the error again, which is, I couldnt visualize the normals with the visualizer.

First I checked if any including or linking of 3rd party libraries were different in the project built manually vs. cmake. But no difference! Then I checked the modules/used DLLs, also no difference.

Finally I saw a PREPROCESSOR definition in the project built with cmake from the tutorial.
After I manually added the same preprocessor defintion in the MSVS2015 project (which was manually built), it made the difference. After that I could visualize the normals like in the tutorial, so no more error.

The preprocessor defintion is called **vtkRenderingCore_INCLUDE** and includes a header file where the define directive
**#define vtkRenderingCore_AUTOINIT 3(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL)**
 is defined for the compiler.
The macro is neccessary to initialze 3 implementation modules (vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL).

As stated by VTK, it is NOT enough to just add this define dirctive in your application code, just where you call vtk, because it has to be done before any vtk header file is included of course.

So to sum up, either you build your project with cmake from the tutorial or you manually add the preprocessor defintion
My preprocessor defintion  for the tutorial cmake project is:
**vtkRenderingCore_INCLUDE="C:/Users/timbe/Desktop/TutorialPCL/build/normal_estimation_using_integral_images/CMakeFiles/vtkRenderingCore_AUTOINIT_vtkInteractionStyle_vtkRenderingFreeType_vtkRenderingOpenGL.h"**

And vtkRenderingCore_AUTOINIT_vtkInteractionStyle_vtkRenderingFreeType_vtkRenderingOpenGL.h has only one statement which is:
**#define vtkRenderingCore_AUTOINIT 3(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL)**

In addition to that there where also defined following preprocessor defintions, which can be necessary as well for other stuff:
vtkRenderingContext2D_AUTOINIT=1(vtkRenderingContextOpenGL)
vtkRenderingVolume_AUTOINIT=1(vtkRenderingVolumeOpenGL)

By the way, I saw you are also doing stuff with kinect and reognition with correspondence grouping in the the lego-lords repository. What is your experience with correspondence grouping from PCL.
I tried to reognize models from scenes from objects in the t-less dataset, see http://cmp.felk.cvut.cz/t-less/
But my correspondence grouping is not working for these scenes, Can you reognize it with your code?
Maybe you have more success.

@SergioRAgostinho Just wanted to let you know, that I found my error in the preprocessor defintion of my project.

**Comment by pclbel** [comment]
see documentation on implementaion modules of vtk
http://www.vtk.org/Wiki/VTK/VTK_6_Migration/Factories_now_require_defines
http://www.vtk.org/Wiki/VTK/Build_System_Migration

**Comment by Lanns** [comment]
hello,could you say more detail about it?I have had the same problem for a long time and I couldt find a useful way to solve it.
i have tried to add 
## #define vtkRenderingCore_AUTOINIT 4(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingFreeTypeOpenGL,vtkRenderingOpenGL)
or
## #define vtkRenderingCore_AUTOINIT 3(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL)
but it doesnt works.
My problem is the same as you that i couldt view the normal as the same error "no override found for vtkActor".I also install pcl by PCL 1.8.0 All-in-one Installer MSVC2015 x64.Must I install it by cmake?i hope i can get your help!Thx very much.

**Comment by pclbel** [comment]
@Lanns Hi, where exactly did you add the line #define vtkRenderingCore_AUTOINIT...?
Just to make sure:
1) As far as I know, it will probably not work, if you have put it into your source code of your application.
2) If you have created a new header file (exlusively for that purpose) having this define statement, make sure that you add the preproccessor statemnet (vtkRenderingCore_INCLUDE="MyCreatedHeaderFile.h") intended to include that newly created header file.

So maybe you can describe more precise where you put your define-statement

**Comment by Lanns** [comment]
@pclbel  Hi,thx your reply.
I think you are right.My project just have "stdafx.h" "targetver.h" "NormalView.cpp"and "stdafx.cpp" and I add the line  #define vtkRenderingCore_AUTOINIT... at the top of "NormalView.cpp" that before the line #include "stdafx.h" just like that:
//
#include <vtkAutoInit.h>
VTK_MODULE_INIT(vtkRenderingOpenGL2)
#define vtkRenderingCore_AUTOINIT 3(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL)

#include "stdafx.h"
#include <pcl/io/io.h>
#include ……
//
So I try to exchange the two like this:
//
#include "stdafx.h"

#include <vtkAutoInit.h>
VTK_MODULE_INIT(vtkRenderingOpenGL2)
#define vtkRenderingCore_AUTOINIT 3(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL)

#include <pcl/io/io.h>
#include ……
//
then I got much error LNK2019 which is about the lib error(maybe?)
In addition,I dont understand what the " (vtkRenderingCore_INCLUDE="MyCreatedHeaderFile.h") " means,could you say more in details?
Im sorry that Im not very good at c++,especially the macro definition,Im just learning about that.
thank you very much!

**Comment by SergioRAgostinho** [comment]
Guys this not really the place for this discussion. Please move the conversation to the mailing list.
Thanks @pclbel for figuring out what was going on.
